File: trunk/geopdes_maxwell/src/op_curlu_curlv_3d.cc
/* Copyright (C) 2009, 2010 Carlo de Falco

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
 
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
*/

#include "geopdes.h"

DEFUN_DLD(op_curlu_curlv_3d, args, nargout,"\n\
OP_CURLU_CURLV_3D: assemble the matrix A = [a(i,j)], a(i,j) = (coeff curl u_j, curl v_i), with a vector-valued curl. \n\
\n\
  mat = op_curlu_curlv_3D (spu, spv, msh, epsilon); \n\
\n\
INPUT: \n\
\n\
  spu:   structure representing the space of trial functions (see sp_bsp_hcurl_3d_phys) \n\
  spv:   structure representing the space of test functions  (see sp_bsp_hcurl_3d_phys) \n\
  msh:   structure containing the domain partition and the quadrature rule (see msh_push_forward_3d) \n\
  coeff: physical parameter \n\
\n\
OUTPUT: \n\
\n\
  mat: assembled matrix \n\
")
{
  
  octave_value_list retval;

  geopdes_mesh  msh (args(2).map_value ());
  geopdes_space spu (args(0).map_value (), msh);
  geopdes_space spv (args(1).map_value (), msh);
  Matrix coeff   = args(3).matrix_value();

  if (!error_state)
    {

      Array <octave_idx_type> I (msh.nel () * spv.nsh_max () * spu.nsh_max (), 0.0);
      Array <octave_idx_type> J (msh.nel () * spv.nsh_max () * spu.nsh_max (), 0.0);
      Array <double> V (msh.nel () * spv.nsh_max () * spu.nsh_max (), 0.0);
            
      SparseMatrix mat;

#pragma omp parallel default (none) shared (msh, spu, spv, I, J, V, coeff)
      {      
      octave_idx_type counter;
#pragma omp for
      for ( octave_idx_type iel=0; iel < msh.nel (); iel++) 
        if (msh.area (iel) > 0.0)
	  {
	    for ( octave_idx_type idof(0); idof < spv.nsh (iel); idof++) 
	      {
	        for (octave_idx_type jdof(0); jdof < spu.nsh (iel); jdof++)
		  {
                    counter = jdof + spu.nsh (iel) * (idof + spv.nsh (iel) * iel);
		    I(counter) = spv.connectivity (idof, iel) - 1;
		    J(counter) = spu.connectivity (jdof,iel)-1;
		    V(counter) = 0.0;
		    for ( octave_idx_type inode(0); inode < msh.nqn (); inode++) 
		      {
		        if (msh.weights (inode, iel) > 0.0)
			  {			  
                            double s = 0.0;                                                    
                            for (octave_idx_type icmp(0); icmp < spu.ncomp (); icmp++)
                              s += spv.shape_function_curls (icmp, inode, idof, iel) * spu.shape_function_curls (icmp, inode, jdof, iel);

			    V(counter) += 
			      msh.jacdet (inode, iel) * 
                              msh.weights (inode, iel) * 
                              coeff (inode, iel) *
                              s;
			  }  
		      } // end for inode		  
//		    if (idof != jdof) // copy upper triangular part to lower
//		      { 
//		        I(counter) = J(counter-1);
//		        J(counter) = I(counter-1);
//		        V(counter) = V(counter-1);
//		        counter++;
//		      } 
		  } // end for jdof
	      } // end for idof
          } else {
          warning_with_id ("geopdes:zero_measure_element", "op_curlu_curlv_3d: element %d has 0 volume", iel);
        }  // end for iel, if area > 0
      } // end omp parallel
      mat = SparseMatrix(V, I, J, spv.ndof (), spu.ndof (), true);
      retval(0) = octave_value (mat);
    } // end if !error_state
  return retval;
}

File: trunk/geopdes_maxwell/inst/operators/op_curlu_curlv_3d.m
% OP_CURLU_CURLV_3D: assemble the matrix A = [a(i,j)], a(i,j) = (coeff curl u_j, curl v_i), with vector-valued curl.
%
%   mat = op_curlu_curlv_3d (spu, spv, msh, epsilon);
%   [rows, cols, values] = op_curlu_curlv_3d (spu, spv, msh, epsilon);
%
% INPUT:
%
%   spu:   structure representing the space of trial functions (see sp_vector_3d_curl_transform/sp_evaluate_col)
%   spv:   structure representing the space of test functions  (see sp_vector_3d_curl_transform/sp_evaluate_col)
%   msh:   structure containing the domain partition and the quadrature rule (see msh_3d/msh_evaluate_col)
%   coeff: physical parameter
%
% OUTPUT:
%
%   mat:    assembled matrix
%   rows:   row indices of the nonzero entries
%   cols:   column indices of the nonzero entries
%   values: values of the nonzero entries
%
% Copyright (C) 2009, 2010 Carlo de Falco
% Copyright (C) 2011 Rafael Vazquez
%
%    This program is free software: you can redistribute it and/or modify
%    it under the terms of the GNU General Public License as published by
%    the Free Software Foundation, either version 3 of the License, or
%    (at your option) any later version.

%    This program is distributed in the hope that it will be useful,
%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%    GNU General Public License for more details.
%
%    You should have received a copy of the GNU General Public License
%    along with this program.  If not, see <http://www.gnu.org/licenses/>.


function varargout = op_curlu_curlv_3d (spu, spv, msh, coeff)
  
  rows = zeros (msh.nel * spu.nsh_max * spv.nsh_max, 1);
  cols = zeros (msh.nel * spu.nsh_max * spv.nsh_max, 1);
  values = zeros (msh.nel * spu.nsh_max * spv.nsh_max, 1);

  ncounter = 0;
  for iel = 1:msh.nel
    if (all (msh.jacdet(:,iel)))
      jacdet_weights = reshape (msh.jacdet(:, iel) .* ...
                       msh.quad_weights(:, iel) .* coeff(:, iel), 1, msh.nqn);

      curlv_iel = reshape (spv.shape_function_curls(:, :, 1:spv.nsh(iel), iel), ...
                            3, msh.nqn, spv.nsh(iel));
      curlu_iel = reshape (spu.shape_function_curls(:, :, 1:spu.nsh(iel), iel), ...
                            3, msh.nqn, spu.nsh(iel));

      curlv_times_jw = bsxfun (@times, jacdet_weights, curlv_iel);
      for idof = 1:spv.nsh(iel)
        rows(ncounter+(1:spu.nsh(iel))) = spv.connectivity(idof, iel);
        cols(ncounter+(1:spu.nsh(iel))) = spu.connectivity(1:spu.nsh(iel), iel);

        aux_val = bsxfun (@times, curlv_times_jw(:,:,idof), curlu_iel);
        values(ncounter+(1:spu.nsh(iel))) = sum (sum (aux_val, 2), 1);
        ncounter = ncounter + spu.nsh(iel);
      end
    else
      warning ('geopdes:jacdet_zero_at_quad_node', 'op_curlu_curlv_3d: singular map in element number %d', iel)
    end
  end

  if (nargout == 1)
    varargout{1} = sparse (rows(1:ncounter), cols(1:ncounter), ...
                           values(1:ncounter), spv.ndof, spu.ndof);
  elseif (nargout == 3)
    varargout{1} = rows(1:ncounter);
    varargout{2} = cols(1:ncounter);
    varargout{3} = values(1:ncounter);
  else
    error ('op_curlu_curlv_3d: wrong number of output arguments')
  end

end

%% COPY OF THE FIRST VERSION OF THE FUNCTION (MORE UNDERSTANDABLE)
% 
% function mat = op_curlu_curlv_3d (spu, spv, msh, coeff)
%   
%   mat = spalloc(spv.ndof, spu.ndof, 1);
%   for iel = 1:msh.nel
%     if (all (msh.jacdet(:,iel)))
%       mat_loc = zeros (spv.nsh(iel), spu.nsh(iel));
%       for idof = 1:spv.nsh(iel)
%         ishc = reshape(spv.shape_function_curls(:,:,idof,iel), 3, []);
%         for jdof = 1:spu.nsh(iel)
%           jshc = reshape (spu.shape_function_curls(:,:,jdof,iel), 3, []);
% % The cycle on the quadrature points is vectorized
%           %for inode = 1:msh.nqn
%             mat_loc(idof, jdof) = mat_loc(idof, jdof) + ...
%               sum (msh.jacdet(:,iel) .* msh.quad_weights(:, iel) .*...
%               sum (ishc .* jshc, 1).' .* coeff (:, iel));
%           %end  
%         end
%       end
%       mat(spv.connectivity(:,iel), spu.connectivity(:,iel)) = ...
%         mat(spv.connectivity(:,iel), spu.connectivity(:,iel)) + mat_loc;
%     else
%       warning ('geopdes:jacdet_zero_at_quad_node', 'op_curlu_curlv_3d: singular map in element number %d', iel)
%     end
%   end
% 
% end